src/relay/debug_printer.cc

```diff
@@ -0,0 +1,261 @@
+    std::vector<Doc> vec;
+    for (size_t i = 0; i < arr.size(); ++i) {
+      vec.push_back(Docify(arr[i]));
+    }
+    return vec;
+  }
+
+  std::vector<Doc> DocifyParamArray(const tvm::Array<Param> &arr) {
+    std::vector<Doc> vec;
+    for (size_t i = 0; i < arr.size(); ++i) {
+      vec.push_back(Docify(arr[i]));
+    }
+    return vec;
+  }
+
+  Doc VisitExpr_(const ConstantNode * c) final {
+    return DocOfStr("some_constant");
+  }
+
+  Doc VisitExpr_(const TupleNode * t) final {
+    return Seq("<", DocifyExprArray(t->fields), ">");
+  }
+
+  Doc VisitExpr_(const VarNode * v) final {
+    return DocOfStr(VarName(GetRef<Var>(v)));
+  }
+
+  Doc VisitExpr_(const GlobalVarNode * g) final {
+    return DocOfStr(g->name_hint);
+  }
+
+  Doc VisitExpr_(const ParamNode * p) final {
+    return TypeAnnotation(Docify(p->var), p->type);
+  }
+
+  Doc VisitExpr_(const FunctionNode * f) final {
+    return Group(TypeAnnotation(Seq("(", DocifyParamArray(f->params), ")"), f->ret_type) + Sep() +
+                 DocOfStr("=>") + Sep() +
+                 Block(indent, "{", Docify(f->body), "}"));
+  }
+
+  Doc VisitExpr_(const CallNode * c) final {
+    auto args = DocifyExprArray(c->args);
+    return Docify(c->op) + Seq("<", DocifyExprArray(c->args), ">");
+  }
+
+  Doc VisitExpr_(const LetNode * l) final {
+    return Group(DocOfStr("let") + Sep() + TypeAnnotation(Docify(l->var), l->value_type) + Sep() +
+                 DocOfStr("=") + Sep() + Docify(l->value) + DocOfStr(";") + Endl() +
+                 Docify(l->body));
+  }
+
+  Doc VisitExpr_(const IfNode * i) final {
+    return Group(DocOfStr("if") + Sep() + Docify(i->cond) + Sep() +
+                 Block(indent, "{", Docify(i->true_branch), "}") + Sep() +
+                 DocOfStr("else") + Sep() +
+                 Block(indent, "{", Docify(i->false_branch), "}"));
+  }
+
+  Doc VisitExpr_(const OpNode * o) final {
+    return DocOfStr(o->name);
+  }
+
+ public:
+  ExprDocifier(const Environment &env) : env(env), td(env) { }
+
+  Doc Docify(const Expr &e) { return (*this)(e); }
+};
+
+RDoc ExprRDoc(const Environment &env, const Expr &expr, WrapWidth width) {
+  return Layout(DocOfExpr(env, expr), width.width);
+}
+
+RDoc TypeRDoc(const Environment &env, const Type &expr, WrapWidth width) {
+  return Layout(DocOfType(env, expr), width.width);
+}
+
+Doc DocOfExpr(const Environment &env, const Expr &expr) {
+  ExprDocifier d(env);
+  return d.Docify(expr);
+}
+
+Doc DocOfType(const Environment &env, const Type &expr) {
+  TypeDocifier d(env);
+  return d.Docify(expr);
+}
+
+std::string PrintExpr(const Environment &env, const Expr &e, WrapWidth width) {
+  std::stringstream ss;
+  ss << ExprRDoc(env, e, width);
+  return ss.str();
+}
+
+std::string PrintType(const Environment &env, const Type &t, WrapWidth width) {
+  std::stringstream ss;
+  ss << TypeRDoc(env, t, width);
+  return ss.str();
+}
+
+TVM_REGISTER_API("relay.debug_printer.print_expr")
```

Comment:
relay.expr._debug_print (which handles both print_expr and print_type, by always print itself)